src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java

```diff
@@ -61,34 +96,177 @@ private void combineShardScores(final ScoreCombinationTechnique scoreCombination
+            if (topDocs.scoreDocs.length != 0) {
+                topFieldDocs.add((TopFieldDocs) topDocs);
+            }
+        }
+        return topFieldDocs;
+    }
+
+    /**
+     *
+      * @param compoundTopDocs top docs that represent on shard
+     * @param combinedNormalizedScoresByDocId docId to normalized scores map
+     * @param sort sort criteria
+     * @return map of docId and sort fields if sorting is enabled.
+     */
+    private Map<Integer, Object[]> getDocIdSortFieldsMap(
+        final CompoundTopDocs compoundTopDocs,
+        final Map<Integer, Float> combinedNormalizedScoresByDocId,
+        final Sort sort
+    ) {
+        // If sort is null then no sort fields present therefore return null.
+        if (sort == null) {
+            return null;
+        }
         // we're merging docs with normalized and combined scores. we need to have only maxHits results
-        List<Integer> sortedDocsIds = new ArrayList<>(combinedNormalizedScoresByDocId.keySet());
-        sortedDocsIds.sort((a, b) -> Float.compare(combinedNormalizedScoresByDocId.get(b), combinedNormalizedScoresByDocId.get(a)));
-        return sortedDocsIds;
+        Map<Integer, Object[]> docIdSortFieldMap = new HashMap<>();
+        List<TopDocs> topFieldDocs = compoundTopDocs.getTopDocs();
+        final boolean isSortByScore = checkIfSortOrderByScore(sort);
+        for (TopDocs topDocs : topFieldDocs) {
+            for (ScoreDoc scoreDoc : topDocs.scoreDocs) {
+                FieldDoc fieldDoc = (FieldDoc) scoreDoc;
+
+                if (docIdSortFieldMap.get(fieldDoc.doc) == null) {
+                    // If sort by score then replace sort field value with normalized score.
+                    if (isSortByScore) {
+                        docIdSortFieldMap.put(fieldDoc.doc, new Object[] { combinedNormalizedScoresByDocId.get(fieldDoc.doc) });
+                    } else {
+                        docIdSortFieldMap.put(fieldDoc.doc, fieldDoc.fields);
+                    }
+                }
+            }
+        }
+        return docIdSortFieldMap;
+    }
+
+    private Collection<Integer> getSortedDocIds(
+        final Map<Integer, Float> combinedNormalizedScoresByDocId,
+        final List<TopFieldDocs> topFieldDocs,
+        final Sort sort
+    ) {
+        // we're merging docs with normalized and combined scores. we need to have only maxHits results
+        if (Objects.isNull(sort)) {
+            List<Integer> sortedDocsIds = new ArrayList<>(combinedNormalizedScoresByDocId.keySet());
+            sortedDocsIds.sort((a, b) -> Float.compare(combinedNormalizedScoresByDocId.get(b), combinedNormalizedScoresByDocId.get(a)));
+            return sortedDocsIds;
+        } else {
+            if (Objects.isNull(topFieldDocs)) {
+                throw new IllegalArgumentException("topFieldDocs cannot be null when sorting is enabled.");
+            }
+            int topN = 0;
```

Comment:
what does topN means?

Reply:
It is a parameter in TopDocs.merge method which is performed few lines below. The merge method as an argument named topN. It signifies how many documents to merge from the TopDocs on the basis of sort criteria. The reason why I kept is as topN so that no one gets confused while debugging the code.